lib/service.go

```diff
@@ -130,126 +169,173 @@ func (m *RelayService) ProposeBlindedBlockV1(r *http.Request, args *SignedBlinde
 	if err != nil {
-		logMethod.WithField("err", err).Error("Could not parse response")
 		return err
 	}
 
-	err = json.Unmarshal(resp.Result, result)
-	if err != nil {
-		logMethod.WithField("err", err).Error("Could not unmarshal response")
-		return err
+	forkchoiceResponses, found := m.store.GetForkchoiceResponse(payloadID.String())
+	if !found {
+		return fmt.Errorf("no ForkChoiceResponses for payloadID %s", payloadID)
 	}
 
-	if result.Transactions != nil {
-		logMethod.WithFields(logrus.Fields{
-			"blockHash": result.BlockHash,
-			"number":    result.Number,
-		}).Info("GetPayloadHeaderV1: calculating tx root from tx list")
-
-		var byteTxs [][]byte
-		for i, otx := range *result.Transactions {
-			var tx types.Transaction
-			bytesTx := common.Hex2Bytes(otx)
-			if err := tx.UnmarshalBinary(bytesTx); err != nil {
-				logMethod.WithFields(logrus.Fields{
-					"err":   err,
-					"tx":    string(bytesTx),
-					"count": i,
-				}).Error("Failed to decode tx")
-				return fmt.Errorf("failed to decode tx %d: %v", i, err)
-			}
-			byteTxs = append(byteTxs, bytesTx)
+	// Call the relay
+	resultC := make(chan *rpcResponseContainer, len(forkchoiceResponses))
+	for relayURL, relayPayloadID := range forkchoiceResponses {
+		go func(url, payloadID string) {
+			res, err := makeRequest(url, "relay_getPayloadHeaderV1", []interface{}{payloadID})
+			resultC <- &rpcResponseContainer{url, err, res}
+		}(relayURL, relayPayloadID)
+	}
+
+	// Collect the responses
+	for i := 0; i < cap(resultC); i++ {
+		res := <-resultC
+
+		// Check for errors
+		if res.err != nil {
+			logMethod.WithFields(logrus.Fields{"error": res.err, "url": res.url}).Warn("error making request to relay")
+			continue
+		}
+		if res.res.Error != nil {
+			logMethod.WithFields(logrus.Fields{"error": res.res.Error, "url": res.url}).Warn("error reply from relay")
+			continue
 		}
 
-		newRootBytes, err := txroot.TransactionsRoot(byteTxs)
+		// Decode response
+		_result := new(ExecutionPayloadWithTxRootV1)
+		err := json.Unmarshal(res.res.Result, _result)
 		if err != nil {
-			logMethod.WithField("err", err).Error("Error calculating tx root")
-			return err
+			logMethod.WithFields(logrus.Fields{"error": err, "data": string(res.res.Result)}).Warn("Could not unmarshal response")
+			continue
 		}
-		newRoot := common.BytesToHash(newRootBytes[:])
 
-		if result.TransactionsRoot != nilHash {
-			if newRoot != result.TransactionsRoot {
-				err := fmt.Errorf("mismatched tx root: %s, %s", newRoot.String(), result.TransactionsRoot.String())
-				logMethod.WithField("err", err).Error("Mismatched tx root")
+		// Skip processing this result if lower fee than previous
+		if result != nil && result.FeeRecipientDiff != nil {
+			if _result.FeeRecipientDiff == nil || _result.FeeRecipientDiff.Cmp(result.FeeRecipientDiff) < 1 {
+				continue
+			}
+		}
+
+		// Use this relay's response as mev-boost response because it's most profitable
+		*result = *_result
```

Comment:
Can `result` be nil here? If not, then I think the check in previous if is a bit confusing.

Reply:
Nope, result is never nil. And yeah, good point, removed nil check in previous if.